simple_network_sim/inference.py

```diff
@@ -0,0 +1,708 @@
+        """ Internal single iteration of ABC-SMC. Sampling particles (from prior or previous accepted ones),
+        until enough particles pass a goodness-of-fit threshold.
+
+        :param smc_step: ABC-SMC iteration number
+        :param prev_particles: List of particles accepted on previous round
+        :param prev_weights: List of particles weights accepted on previous round
+        :return: List of particles and weights accepted at current ABC-SMC round
+        """
+        t0 = time.time()
+
+        particles = []
+        weights = []
+        distances = []
+
+        particles_accepted = 0
+        particles_simulated = 0
+        while particles_accepted < self.n_particles:
+
+            if smc_step == 0:
+                particle = Particle.generate_from_priors(self)
+            else:
+                particle = Particle.resample_and_perturbate(prev_particles, prev_weights, self.random_state)
+
+            result = self.run_model(particle)
+            distance = self.compute_distance(result)
+
+            if distance <= self.threshold:
+                logger.info("Particle accepted with distance %d", distance)
+                weight = ABCSMC.compute_weight(smc_step, prev_particles, prev_weights, particle)
+                particles.append(particle)
+                weights.append(weight)
+                distances.append(distance)
+                particles_accepted += 1
+
+            particles_simulated += 1
+
+        logger.info("Particles accepted %d/%d", particles_accepted, particles_simulated)
+        self.add_iteration_statistics(smc_step, particles, particles_simulated, distances, t0)
+
+        return particles, weights, distances
+
+    def run_model(self, particle: Particle) -> pd.DataFrame:
+        """ Run models using current particle as parameters.
+
+        :param particle: Particle under consideration
+        :return: Model run results for current particle
+        """
+        network = ss.createNetworkOfPopulation(
+            self.compartment_transition_table,
+            self.population_table,
+            self.commutes_table,
+            self.mixing_matrix_table,
+            self.infectious_states,
+            particle.inferred_variables["infection-probability"].value,
+            particle.inferred_variables["initial-infections"].value,
+            self.trials,
+            self.start_end_date,
+            self.movement_multipliers_table,
+            self.stochastic_mode,
+            self.random_seed
+        )
+        results = sm.runSimulation(network)
+        aggregated = sm.aggregateResults(results)
+        return aggregated.output
+
+    def compute_distance(self, result: pd.DataFrame) -> float:
+        """ Computes distance between target and model run with current particle.
+        The distance is the root mean squared distance between model and historical
+        death rates:
+
+        sqrt(mean((Target_death_rate - Model_death_rate)**2))
```

Comment:
This ABCSMC procedure is pretty specific to the current model and choice of fitting parameters (e.g the mention of deaths).  This is unavoidable at some level, but perhaps worth mentioning in the class description or denoting some other way?

Reply:
So from what I've read, using root mean square of differences for all times is standard in epidemiology and for dynamical systems in general (at least this is what https://royalsocietypublishing.org/doi/pdf/10.1098/rsif.2008.0172 is saying).

But indeed the fact that the time series of focus for us are the deaths through time is specific to this model, so we should probably abstract that away, especially given we might add hospitalizations too. I'll make it a bit more generic and add to comments.